### PR TITLE
Disable wasm in gen_struct_info.py

### PR DESCRIPTION
**Pull request by kripken**
When boostrapping `struct_info.compiled.json`, there is no need for wasm (we are just compiling small programs to detect C constants from headers). And in fact it builds the binaryen port just for that, which confuses `sanity.test_emcc_ports`.

### CODE REVIEW

**Comment by jgravelle-google** [comment]
Also this gen_struct_info change breaks the wasm backend path :/
https://logs.chromium.org/v/?s=chromium%2Fbb%2Fclient.wasm.llvm%2Flinux%2F31841%2F%2B%2Frecipes%2Fsteps%2Femscripten__emwasm_%2F0%2Fstdout
```
assert shared.Settings.BINARYEN_ROOT, 'need BINARYEN_ROOT config set so we can use Binaryen s2wasm on the backend output'
AssertionError: need BINARYEN_ROOT config set so we can use Binaryen s2wasm on the backend output
```

I have no idea why that's not set there. Any ideas?

**Comment by jgravelle-google** [comment]
Er, for clarity: 60854ef434231ad6616acd3bfb72389ca714179b is the commit that breaks it.

**Comment by kripken** [comment]
Hmm... actually, are you sure it was that commit and not this one that broke it? It seems like this "fix" is bad for the wasm backend because it tries to build without wasm to bootstrap, and fails (sorry about that!). The error you linked to looks like that's what's happening there, BINARYEN_ROOT is not set.

Or did the previous commit break it in one way, then this commit broke it in another? If so, would be good to see the error for that.

In other words, I think we should revert this "fix" and do something better, but first I'd like to know if there's also a problem with the previous commit, and if so what, as the something better may depend on that.

**Comment by jgravelle-google** [comment]
Positive - the waterfall breakage started before this change landed. Also I tested the "wasm-by-default" patch first, because that seemed most likely to have broken it, and it didn't.
Applying this patch locally it fails in the same way, so we probably don't need to revert it.

For wasm backend `-s WASM=0` is ignored, because we can only build with wasm.

So the problem is there's no BINARYEN_ROOT, which is bewildering because there is in the `.emscripten` I'm using has it set.

... ok looking more into this, `-s WASM=0` will be a problem because we set settings.BINARYEN_ROOT to shared.BINARYEN_ROOT only if settings.WASM, in ports/binaryen.py. Only that doesn't matter because we don't actually call that function when we assert, we just check settings.BINARYEN_ROOT without ever looking at the port when calling via gen_struct_info.py.

**Comment by jgravelle-google** [comment]
Ok so the problem is we set settings.BINARYEN_ROOT from shared.BINARYEN_ROOT but only if we're loading the binaryen port, which we don't do when BOOTSTRAPPING_STRUCT_INFO.

A nasty hack that gets around that is to say
```python
    if not shared.Settings.BINARYEN_ROOT:
      shared.Settings.BINARYEN_ROOT = shared.BINARYEN_ROOT
```
in the wasm_backend emscripten.py path.

**Comment by jgravelle-google** [comment]
Seems to be more like:
1. Building generated_struct_info.json with wasm backend has never worked
1. generated_struct_info.json used to be checked in to the tree so that didn't matter
1. https://github.com/kripken/emscripten/commit/60854ef434231ad6616acd3bfb72389ca714179b made it so that each target builds their own generated_struct_info.json (which is a good idea - I'm amazed things weren't broken by asm.js and wasm_backend disagreeing about how that should look?)
1. #6521 is probably a reasonable simplification but doesn't fix the wasm backend at all

Trying #6521 locally it doesn't fix the problem, and I wouldn't expect it to. The problem isn't that we need to fetch the binaryen port, the problem is that we need to call `needed` to set `settings.BINARYEN_ROOT`, because that's what the rest of the pipeline depends on.

**Comment by kripken** [comment]
That seems to make sense, but we do have that `EMCC_WASM_BACKEND_BINARYEN` code in `gen_struct_info` (being removed in #6521 ... ;) ) which suggests we made it work in the wasm backend at some point - is that really false?

> we need to call `needed`

What is `needed`?

If the issue is `BINARYEN_ROOT` is not set early enough for gen_struct_info, then we may need to call it in there, or something like that, yeah.

**Comment by jgravelle-google** [comment]
Whether we need to `get` the port or not: https://github.com/kripken/emscripten/blob/6e5818017d1b2e09e9f7ad22a32e9a191f6f9a3b/tools/ports/binaryen.py#L5-L16

It just has the side effect of setting `settings.BINARYEN_ROOT`. The problem is that this is the **only** place that sets `settings.BINARYEN_ROOT`. This actually kind of makes sense, because most people won't care and will get binaryen via ports, whereas as toolchain authors we set it in `.emscripten`.

The problem with `gen_struct_info.py` is that we set `BOOTSTRAPPING_STRUCT_INFO=1`, which makes us ignore ports:
https://github.com/kripken/emscripten/blob/6e5818017d1b2e09e9f7ad22a32e9a191f6f9a3b/emcc.py#L1528-L1535

https://github.com/kripken/emscripten/pull/6522 seems to fix this locally. In theory this should be reasonable for end-users, because we've already fetched the port by this point.

**Comment by kripken** [comment]
Interesting, let's move the discussion to your PR.
